sdk/core/core/inc/az_span.h

```diff
@@ -71,209 +50,69 @@ AZ_INLINE az_result_byte az_const_span_get(az_const_span const span, size_t cons
 }
 
-AZ_INLINE bool az_span_is_overlap(az_span const a, az_span const b) {
-  return az_const_span_is_overlap(az_span_to_const_span(a), az_span_to_const_span(b));
-}
-
-AZ_INLINE az_result
-az_span_copy(az_span const buffer, az_const_span const src, az_span * const out_result) {
-  AZ_CONTRACT_ARG_NOT_NULL(out_result);
-  if (!az_span_is_valid(buffer) || !az_const_span_is_valid(src)) {
-    return AZ_ERROR_ARG;
-  }
-
-  if (az_const_span_is_overlap(az_span_to_const_span(buffer), src)) {
-    return AZ_ERROR_ARG;
-  }
-
-  if (buffer.size < src.size) {
-    return AZ_ERROR_BUFFER_OVERFLOW;
-  }
-
-  if (!az_const_span_is_empty(src)) {
-    memcpy((void *)buffer.begin, (void const *)src.begin, src.size);
-  }
-
-  out_result->begin = buffer.begin;
-  out_result->size = src.size;
-
-  return AZ_OK;
-}
-
-AZ_INLINE az_span az_span_drop(az_span const span, size_t const n) {
-  if (span.size <= n) {
-    return (az_span){ .begin = NULL, .size = 0 };
-  }
-  return (az_span){ .begin = span.begin + n, .size = span.size - n };
-}
-
-AZ_INLINE az_span az_span_take(az_span const span, size_t n) {
-  if (span.size <= n) {
-    return span;
-  }
-  return (az_span){ .begin = span.begin, .size = n };
-}
-
-AZ_INLINE az_result az_span_set(az_span const span, uint8_t const fill) {
-  if (!az_span_is_valid(span)) {
-    return AZ_ERROR_ARG;
-  }
-
-  if (!az_span_is_empty(span)) {
-    memset(span.begin, fill, span.size);
-  }
-
-  return AZ_OK;
-}
-
-AZ_INLINE az_result az_span_swap(az_span const a, az_span const b) {
-  if (!az_span_is_valid(a) || !az_span_is_valid(b)) {
-    return AZ_ERROR_ARG;
-  }
-
-  if (a.size != b.size || az_span_is_overlap(a, b)) {
-    return AZ_ERROR_ARG;
-  }
-
-  if (!az_span_is_empty(a)) {
-    uint8_t * a_ptr = a.begin;
-    uint8_t * b_ptr = b.begin;
-    uint8_t const * const a_end = a.begin + a.size;
-    do {
-      uint8_t const old_a = *a_ptr;
-      *a_ptr = *b_ptr;
-      *b_ptr = old_a;
-
-      ++a_ptr;
-      ++b_ptr;
-    } while (a_ptr != a_end);
-  }
-
-  return AZ_OK;
-}
-
-AZ_INLINE az_result
-az_span_to_c_str(az_span const buffer, az_const_span const src, az_span * const out_result) {
-  AZ_CONTRACT_ARG_NOT_NULL(out_result);
-
-  if (!az_const_span_is_valid(src)) {
-    return AZ_ERROR_ARG;
-  }
-
-  AZ_CONTRACT_ARG_NOT_NULL(buffer.begin);
-  if (buffer.size < src.size + 1) {
-    return AZ_ERROR_BUFFER_OVERFLOW;
-  }
-
-  if (!az_const_span_is_empty(src)) {
-    memmove((void *)buffer.begin, (void const *)src.begin, src.size);
-  }
-
-  buffer.begin[src.size] = '\0';
-
-  out_result->begin = buffer.begin;
-  out_result->size = src.size + 1;
-
-  return AZ_OK;
-}
-
-AZ_NODISCARD az_result az_span_replace(
-    az_span const buffer,
-    az_const_span const src,
-    uint8_t (*const func)(uint8_t const),
-    az_span * const out_result);
-
 #define AZ_ARRAY_SIZE(ARRAY) (sizeof(ARRAY) / sizeof(*ARRAY))
 
-#define AZ_SPAN(ARRAY) \
+#define AZ_SPAN_FROM_ARRAY(ARRAY) \
```

Comment:
this is asking for bugs because people will use it on arrays that have decayed to pointers.

Reply:
you may be able to do this with an inline function and the whole [static] c99 thing, but I'm not sure.

Reply:
yes. Do you have a better idea how we can know the size of the array?

Reply:
@barcharcraz It's easy in C++ but I have no idea how to do it in C99 safely.

Reply:
Reopen it if you have a solution